cluster-autoscaler/core/utils.go

```diff
@@ -112,34 +112,41 @@ func (podMap podSchedulableMap) set(pod *apiv1.Pod, schedulable bool) {
 // It takes into account pods that are bound to node and will be scheduled after lower priority pod preemption.
 func FilterOutSchedulable(unschedulableCandidates []*apiv1.Pod, nodes []*apiv1.Node, allScheduled []*apiv1.Pod, podsWaitingForLowerPriorityPreemption []*apiv1.Pod,
 	predicateChecker *simulator.PredicateChecker, expendablePodsPriorityCutoff int) []*apiv1.Pod {
-
 	unschedulablePods := []*apiv1.Pod{}
 	nonExpendableScheduled := FilterOutExpendablePods(allScheduled, expendablePodsPriorityCutoff)
 	nodeNameToNodeInfo := scheduler_util.CreateNodeNameToInfoMap(append(nonExpendableScheduled, podsWaitingForLowerPriorityPreemption...), nodes)
 	podSchedulable := make(podSchedulableMap)
-
 	loggingQuota := glogx.PodsLoggingQuota()
 
 	for _, pod := range unschedulableCandidates {
-		if schedulable, found := podSchedulable.get(pod); found {
-			if !schedulable {
+		err, found := podSchedulable.get(pod)
+		// Try to get result from cache.
+		if found {
+			if err != nil {
 				unschedulablePods = append(unschedulablePods, pod)
 			} else {
 				glogx.V(4).UpTo(loggingQuota).Infof("Pod %s marked as unschedulable can be scheduled (based on simulation run for other pod owned by the same controller). Ignoring in scale up.", pod.Name)
 			}
-			continue
 		}
-		if nodeName, err := predicateChecker.FitsAny(pod, nodeNameToNodeInfo); err == nil {
-			glogx.V(4).UpTo(loggingQuota).Infof("Pod %s marked as unschedulable can be scheduled on %s. Ignoring in scale up.", pod.Name, nodeName)
-			podSchedulable.set(pod, true)
-		} else {
-			unschedulablePods = append(unschedulablePods, pod)
-			podSchedulable.set(pod, false)
+		// Not found in cache, have to run the predicates.
+		if !found {
+			nodeName, err := predicateChecker.FitsAny(pod, nodeNameToNodeInfo)
+			// err returned from FitsAny isn't a PredicateError.
+			// Hello, ugly hack. I wish you weren't here.
+			var predicateError *simulator.PredicateError
+			if err != nil {
+				predicateError = simulator.NewPredicateError("FitsAny", err, nil)
+			}
+			podSchedulable.set(pod, predicateError)
+			if err != nil {
```

Comment:
nit: I would merge this with identical if 3 lines above and just push podSchedulable.set to the end of the function.

Reply:
Done.